extensions/amp-analytics/0.1/platform-variables.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getService} from '../../../src/service';
+import {documentInfoFor} from '../../../src/document-info';
+import {parseUrl, removeFragment} from '../../../src/url';
+
+export function platformVariablesFor(window) {
+  return getService(window, 'amp-analytics-platform-variables', () => {
+    return platformVariables(window);
+  });
+}
+
+export function platformVariablesExpressionFor(window) {
+  const vars = platformVariablesFor(window);
+  return new RegExp('\{(' + Object.keys(vars).join('|') + ')\}', 'g');
+}
+
+function platformVariables(win) {
+  return {
+    /**
+     * Returns a random value for cache busters.
+     */
+    'random': () => {
```

Comment:
Although url-replacements is extensible, a design where tokens are simply uppercase strings will run into problems if one token identifier contains a substring of any other token identifier. Since we allow publishers to specify their own tokens (in the amp-analytics config / trigger config) we're going to run into that case eventually.

I haven't thought this through very deeply, but on the surface it seems easier to implement a system where tokens can be identified in a template string without knowledge of all possible token identifiers. E.g., if tokens are contained within curly braces, you can do something like this:

  templateString.replace(/{([^}]*)}/g, (m) => { return dataModel.get(m) });

(vs. the concatenation of all identifiers approach in url-replacement.)